bazel_codegen/lib/src/arg_parser.dart

```diff
@@ -0,0 +1,130 @@
+          'package name.')
+  ..addOption(_srcsParam,
+      help: 'Path to a file containing all files to generate code for. '
+          'Each line in this file is the path to a source to generate for. '
+          'These are expected to be relative to CWD.');
+
+Map<String, Level> _optionToLogLevel = {
+  'fine': Level.FINE,
+  'info': Level.INFO,
+  'warning': Level.WARNING,
+  'error': Level.SEVERE,
+};
+
+/// Parsed arguments for code generator binaries.
+class Options {
+  final List<String> rootDirs;
+  final String packagePath;
+  final String outDir;
+  final Level logLevel;
+  final String logPath;
+  final String inputExtension;
+  final List<String> outputExtensions;
+  final String packageMapPath;
+  final String srcsPath;
+  final bool help;
+  final bool isWorker;
+  final bool useSummaries;
+  final List<String> additionalArgs;
+
+  Options._(
+      this.rootDirs,
+      this.packagePath,
+      this.outDir,
+      this.logPath,
+      this.inputExtension,
+      this.outputExtensions,
+      this.packageMapPath,
+      this.srcsPath,
+      this.help,
+      this.logLevel,
+      {this.isWorker,
+      this.useSummaries: true,
+      this.additionalArgs});
+
+  factory Options.parse(List<String> args, {bool isWorker}) {
+    // When not running as a worker, but that mode is supported, then we get
+    // just this arg which points at a file containing the arguments.
+    if (args.length == 1 && args.first.startsWith('@')) {
+      args = new File(args.first.substring(1)).readAsLinesSync();
+    }
+
+    final argResults = _argParser.parse(args);
+
+    final rootDirs = _checkNotNull(argResults, _rootDirParam) as List<String>;
+    final packagePath = _checkNotNull(argResults, _packagePathParam);
+    final outDir = _checkNotNull(argResults, _outParam);
+    final logLevel =
+        _optionToLogLevel[_checkNotNull(argResults, _logLevelParam)];
+
+    final logPath = _checkNotNull(argResults, _logPathParam);
+    final inputExtension = _checkNotNull(argResults, _inExtension);
+    final outputExtensions =
+        _checkNotNull(argResults, _outExtension) as List<String>;
+
+    final packageMapPath = _checkNotNull(argResults, _packageMapParam);
+    final srcsPath = _checkNotNull(argResults, _srcsParam);
+    final help = argResults[_helpParam];
+    final useSummaries = argResults[_summariesParam];
+
+    return new Options._(rootDirs, packagePath, outDir, logPath, inputExtension,
+        outputExtensions, packageMapPath, srcsPath, help, logLevel,
+        additionalArgs: argResults.rest,
+        isWorker: isWorker,
+        useSummaries: useSummaries);
+  }
+
+  void printUsage() {
+    print('Usage: dart ${Platform.script.pathSegments.last} '
+        '<options(s)> <Additional args for generator>');
+    print('All options are required');
+    print('Options:\n${_argParser.usage}');
+  }
+}
+
+dynamic _checkNotNull(ArgResults results, String param) {
```

Comment:
Lets rename this to `_getRequiredArg` or something along those lines, and maybe make it a generic method to avoid explicit casts?